.config/doom/config.org
#+BEGIN_SRC emacs-lisp
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-

;; Place your private configuration here! Remember, you do not need to run 'doom
;; sync' after modifying this file!


;; Some functionality uses this to identify you, e.g. GPG configuration, email
;; clients, file templates and snippets.
(setq user-full-name "Pavel Atanasov"
      user-mail-address "pavel.atanasov2001@gmail.com")

;; Doom exposes five (optional) variables for controlling fonts in Doom. Here
;; are the three important ones:
;;
;; + `doom-font'
;; + `doom-variable-pitch-font'
;; + `doom-big-font' -- used for `doom-big-font-mode'; use this for
;;   presentations or streaming.
;;
;; They all accept either a font-spec, font string ("Input Mono-12"), or xlfd
;; font string. You generally only need these two:
;; (setq doom-font (font-spec :family "monospace" :size 12 :weight 'semi-light)
;;       doom-variable-pitch-font (font-spec :family "sans" :size 13))

;; There are two ways to load a theme. Both asume the theme is installed and
;; available. You can either set `doom-theme' or manually load a theme with the
;; `load-theme' function. This is the default:
(setq doom-theme 'doom-molokai)

;; If you use `org' and don't want your org files in the default location below,
;; change `org-directory'. It must be set before org loads!
(setq org-directory "~/org/")

;; This determines the style of line numbers in effect. If set to `nil', line
;; numbers are disabled. For relative line numbers, set this to `relative'.
(setq display-line-numbers-type 'relative)


;; Here are some additional functions/macros that could help you configure Doom:
;;
;; - `load!' for loading external *.el files relative to this one
;; - `use-package!' for configuring packages
;; - `after!' for running code after a package has loaded
;; - `add-load-path!' for adding directories to the `load-path', relative to
;;   this file. Emacs searches the `load-path' when you load packages with
;;   `require' or `use-package'.
;; - `map!' for binding new keys
;;
;; To get information about any of these functions/macros, move the cursor over
;; the highlighted symbol at press 'K' (non-evil users must press 'C-c c k').
;; This will open documentation for it, including demos of how they are used.
;;
;; You can also try 'gd' (or 'C-c c d') to jump to their definition and see how
;; they are implemented.

(prefer-coding-system           'utf-8)
(set-default-coding-systems     'utf-8)
(set-terminal-coding-system     'utf-8)
(set-keyboard-coding-system     'utf-8)
(setq default-buffer-file-encoding-system 'utf-8)
(setq x-select-request-type '(UTF8_STRING COMPOUND_TEXT TEXT STRING))

(setq-default indent-tabs-mode nil)
(setq-default tab-width 4)
(setq c-set-style "k&r")
(setq c-basic-offset 4)

(use-package! tree-sitter
  :config
  (require 'tree-sitter-langs)
  (global-tree-sitter-mode)
  (add-hook 'tree-sitter-after-on-hook #'tree-sitter-hl-mode)
  (pushnew! tree-sitter-major-mode-language-alist
            '(scss-mode . css)))

(setq redisplay-dont-pause t
  scroll-margin 1
  scroll-step 1
  scroll-conservatively 10000
  scroll-preserve-screen-position 1)

;; prepare the arguments
(setq auth-sources '("~/.authinfo.gpg"))
(setq dotfiles-git-dir (concat "--git-dir=" (expand-file-name "~/dotfiles")))
(setq dotfiles-work-tree (concat "--work-tree=" (expand-file-name "~")))

;; function to start magit on dotfiles
(defun dotfiles-magit-status ()
  (interactive)
  (add-to-list 'magit-git-global-arguments dotfiles-git-dir)
  (add-to-list 'magit-git-global-arguments dotfiles-work-tree)
  (call-interactively 'magit-status))
(map! :leader
       (:prefix ("g" . "git")
        :desc "Dotfiles status" "d" #'dotfiles-magit-status))

;; wrapper to remove additional args before starting magit
(defun magit-status-with-removed-dotfiles-args ()
  (interactive)
  (setq magit-git-global-arguments (remove dotfiles-git-dir magit-git-global-arguments))
  (setq magit-git-global-arguments (remove dotfiles-work-tree magit-git-global-arguments))
  (call-interactively 'magit-status))
;; redirect global magit hotkey to our wrapper
(map! :leader
       (:prefix ("g" . "git")
        :desc "Magit status" "g" #'magit-status-with-removed-dotfiles-args))


(defun reo101/org-mode-setup ()
  (org-indent-mode)
  (visual-line-mode 1))

(use-package! org
  :hook (org-mode . reo101/org-mode-setup)
  :custom
  (org-ellipsis " ▼")
  (org-hide-emphasis-markers t)
  (org-src-window-setup 'current-window))

;; (defun reo101/org-mode-visual-fill ()
;;  (setq visual-fill-column-width 125
;;        visual-fill-column-center-text t)
;;  (visual-fill-column-mode 1))
;;
;; (use-package! visual-fill-column
;;   :diminish visual-line-mode
;;   :hook (org-mode . reo101/org-mode-visual-fill))

(defun reo101/org-mode-big-headings ()
  (dolist (face '((org-level-1 . 1.4)
              (org-level-2 . 1.35)
              (org-level-3 . 1.30)
              (org-level-4 . 1.25)
              (org-level-5 . 1.20)
              (org-level-6 . 1.15)
              (org-level-7 . 1.10)
              (org-level-8 . 1.05)))
  (set-face-attribute (car face) nil ;; :font "Cantarell"
                      :weight 'regular :height (cdr face))))

(use-package! org-faces
  :hook (org-mode . reo101/org-mode-big-headings))

(use-package! org-bullets
  :hook (org-mode . org-bullets-mode)
  :custom (org-bullets-bullet-list '("◉" "○" "●" "○" "●" "○" "●")))

;; (setq
;;     org-superstar-headline-bullets-list '("◉" "○" "●" "○" "●" "○" "●")
;; )

(defun reo101/org-mode-checkbox-todo ()
  "Switch header TODO state to DONE when all checkboxes are ticked, to TODO otherwise"
  (let ((todo-state (org-get-todo-state)) beg end)
    (unless (not todo-state)
      (save-excursion
    (org-back-to-heading t)
    (setq beg (point))
    (end-of-line)
    (setq end (point))
    (goto-char beg)
    (if (re-search-forward "\\[\\([0-9]*%\\)\\]\\|\\[\\([0-9]*\\)/\\([0-9]*\\)\\]"
                   end t)
        (if (match-end 1)
        (if (equal (match-string 1) "100%")
            (unless (string-equal todo-state "DONE")
              (org-todo 'done))
          (unless (string-equal todo-state "TODO")
                (org-todo 'todo)))
          (if (and (> (match-end 2) (match-beginning 2))
               (equal (match-string 2) (match-string 3)))
          (unless (string-equal todo-state "DONE")
            (org-todo 'done))
        (unless (string-equal todo-state "TODO")
          (org-todo 'todo)))))))))

(add-hook 'org-checkbox-statistics-hook 'reo101/org-checkbox-todo)

(font-lock-add-keywords 'org-mode
            '(("^ *\\([-]\\) "
               (0 (prog1 () (compose-region (match-beginning 1)
                                          (match-end 1) "•"))))))

(setq-default prettify-symbols-alist '(("#+BEGIN_SRC" . "†")
                                       ("#+END_SRC" . "†")
                                       ("#+begin_src" . "†")
                                       ("#+end_src" . "†")
                                       (">=" . "≥")
                                       ("=>" . "⇨")))
(setq prettify-symbols-unprettify-at-point 'right-edge)
(add-hook 'org-mode-hook 'prettify-symbols-mode)

(set-frame-parameter (selected-frame) 'alpha '(95 . 85))
(add-to-list 'default-frame-alist '(alpha . (95 . 85)))

(defun reo101/toggle-transparency ()
(interactive)
(let ((alpha (frame-parameter nil 'alpha)))
    (set-frame-parameter
    nil 'alpha
    (if (eql (cond ((numberp alpha) alpha)
                    ((numberp (cdr alpha)) (cdr alpha))
                    ;; Also handle undocumented (<active> <inactive>) form.
                    ((numberp (cadr alpha)) (cadr alpha)))
            100)
        '(95 . 85) '(100 . 100)))))
(map! :leader
       (:prefix ("t" . "toggle")
        :desc "Transparency" "t" #'reo101/toggle-transparency))

(add-hook 'activate-mark-hook 'rainbow-delimiters-mode-disable)
(add-hook 'deactivate-mark-hook 'rainbow-delimiters-mode-enable)

(defun reo101/org-mode-show-current-heading-tidily ()
  (interactive)
  "Show next entry, keeping other entries closed."
  (if (save-excursion (end-of-line) (outline-invisible-p))
      (progn (org-show-entry) (show-children))
    (outline-back-to-heading)
    (unless (and (bolp) (org-on-heading-p))
      (org-up-heading-safe)
      (hide-subtree)
      (error "Boundary reached"))
    (org-overview)
    (org-reveal t)
    (org-show-entry)
    (show-children)))

(use-package! highlight-indent-guides
  :commands highlight-indent-guides-mode
  :hook (prog-mode . highlight-indent-guides-mode)
  :config
  (setq highlight-indent-guides-method 'character
        highlight-indent-guides-character ?\»
        highlight-indent-guides-delay 0.01
        highlight-indent-guides-responsive 'top
        highlight-indent-guides-auto-enabled nil)) ;nil

(use-package! wc-mode
  :hook (org-mode . wc-mode)
  :config
  (setq wc-modeline-format "[Words: %tw, Chars: %tc]"))

;; (map! :leader
;;       (:prefix-map ("b" . "buffer")
;;        (:prefix ("j" . "journal")
;;         :desc "New journal entry" "j" #'org-journal-new-entry
;;         :desc "Search journal entry" "s" #'org-journal-search)))

(map! :leader
      (:prefix-map ("b" . "buffer")
       (:prefix ("w" . "words")
        :desc "Word count (words and characters)" "a" #'wc-count)))

; Spc-b-w

;; (use-package! lsp-mode
;;   :hook ((prog-mode . lsp-deferred))
;;   :commands (lsp lsd-deferred)
;;   :config
;;   (progn
;;     (lsp-register-client
;;      (make-lsp-client :new-connection (lsp-tramp-connection "clangd")
;;                       :major-modes '(c-mode c++-mode)
;;                       :remote? t
;;                       :server-id 'clangd-remote))))

(defun reo101/lsp-mode-setup ()
  (setq lsp-headerline-breadcrumb-segments '(path-up-to-project file symbols))
  (lsp-headerline-breadcrumb-mode))

(use-package! lsp-mode
  :diminish
  :commands (lsp lsp-deferred)
  ;; :bind-keymap ("C-c l" . lsp-command-map)
  :custom
  (lsp-log-io t)
  ;; (lsp-keymap-prefix "C-c l")
  (lsp-register-client
  (make-lsp-client :new-connection (lsp-stdio-connection "intelephense")
                   :major-modes '(c++-mode)
                   :server-id 'intelephense))
  :hook
  ;; (erlang-mode . lsp)
  (c++-mode . lsp)
  ;; (latex-mode . lsp)
  ;; (latex-math-mode . lsp)
  ;; (php-mode . lsp)
  (lsp-mode . reo101/lsp-mode-setup)
  (lsp-mode . lsp-enable-which-key-integration))

(require 'lsp)
(require 'lsp-haskell)
;; Hooks so haskell and literate haskell major modes trigger LSP setup
(add-hook 'haskell-mode-hook #'lsp)
(add-hook 'haskell-literate-mode-hook #'lsp)

(use-package! lsp-ui
  :hook (lsp-mode . lsp-ui-mode)
  :commands lsp-ui
  :custom
  (lsp-ui-sideline-enable t)
  (lsp-ui-doc-enable t)
  (lsp-ui-doc-position 'bottom))

;; (use-package! lsp-origami
;;   :hook
;;   (erlang-mode . origami-mode)
;;   (origami-mode . lsp-origami-mode))

(use-package! lsp-ivy :commands lsp-ivy-workspace-symbol)

; (use-package! origami
;   :bind-keymap ("C-c o" . origami-mode-map)
;   :bind (:map origami-mode-map
;               ("C-c o o" . origami-open-node)
;               ("C-c o O" . origami-open-node-recursively)
;               ("C-c o c" . origami-close-node)
;               ("C-c o C" . origami-close-node-recursively)
;               ("C-c o a" . origami-toggle-node)
;               ("C-c o A" . origami-recursively-toggle-node)
;               ("C-c o R" . origami-open-all-nodes)
;               ("C-c o M" . origami-close-all-nodes)
;               ("C-c o v" . origami-show-only-node)
;               ("C-c o k" . origami-previous-fold)
;               ("C-c o j" . origami-forward-fold)
;               ("C-c o x" . origami-reset)))

(use-package! company
  :after lsp-mode
  :diminish
  :custom
  (company-idle-delay 0)
  (company-minimum-prefix-length 1)
  :bind (:map company-active-map
              ("M-n" . nil)
              ("M-p" . nil)
              ("C-n" . company-select-next)
              ("C-p" . company-select-previous))
  :hook (lsp-mode . company-mode))

(use-package! company-box
  :diminish
  :hook (company-mode . company-box-mode))

; (defvar my-term-shell "/bin/zsh")
; (defadvice ansi-term (before force-bash)
;   (interactive (list my-term-shell)))
; (ad-activate 'ansi-term)
; (global-set-key (kbd "<M-return>") 'ansi-term)

(use-package! sudo-edit
  :bind ("C-c s" . sudo-edit))

; (use-package! ivy
;   :diminish
;   :bind (:map ivy-minibuffer-map
;               ("TAB" . ivy-alt-done))
;   :config (ivy-mode 1))
; (use-package! ivy-rich
;   :init (ivy-rich-mode 1))

; (set-frame-parameter (selected-frame) 'alpha '(85 . 50))
; (add-to-list 'default-frame-alist '(alpha . (85 . 50)))

(use-package! rainbow-delimiters
  :hook (prog-mode . rainbow-delimiters-mode))

;; (prettify-utils-add-hook tex-mode
;;                          ("\\smallo" "o"))

(use-package! org-super-agenda
  :after org-agenda
  :init
  (setq org-super-agenda-groups '((:name "Today"
                                         :time-grid t
                                         :scheduled today)
                                  (:name "Due today"
                                         :deadline today)
                                  (:name "Important"
                                         :priority "A")
                                  (:name "Overdue"
                                         :deadline past)
                                  (:name "Due soon"
                                         :deadline future)
                                  (:name "Big Outcomes"
                                         :tag "bo")))
  :config
  (org-super-agenda-mode))
#+END_SRC


.config/doom/config.el
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-

;; Place your private configuration here! Remember, you do not need to run 'doom
;; sync' after modifying this file!


;; Some functionality uses this to identify you, e.g. GPG configuration, email
;; clients, file templates and snippets.
(setq user-full-name "Pavel Atanasov"
      user-mail-address "pavel.atanasov2001@gmail.com")

;; Doom exposes five (optional) variables for controlling fonts in Doom. Here
;; are the three important ones:
;;
;; + `doom-font'
;; + `doom-variable-pitch-font'
;; + `doom-big-font' -- used for `doom-big-font-mode'; use this for
;;   presentations or streaming.
;;
;; They all accept either a font-spec, font string ("Input Mono-12"), or xlfd
;; font string. You generally only need these two:
;; (setq doom-font (font-spec :family "monospace" :size 12 :weight 'semi-light)
;;       doom-variable-pitch-font (font-spec :family "sans" :size 13))

;; There are two ways to load a theme. Both asume the theme is installed and
;; available. You can either set `doom-theme' or manually load a theme with the
;; `load-theme' function. This is the default:
(setq doom-theme 'doom-dracula)

;; If you use `org' and don't want your org files in the default location below,
;; change `org-directory'. It must be set before org loads!
(setq org-directory "~/org/")

;; This determines the style of line numbers in effect. If set to `nil', line
;; numbers are disabled. For relative line numbers, set this to `relative'.
(setq display-line-numbers-type 'relative)


;; Here are some additional functions/macros that could help you configure Doom:
;;
;; - `load!' for loading external *.el files relative to this one
;; - `use-package!' for configuring packages
;; - `after!' for running code after a package has loaded
;; - `add-load-path!' for adding directories to the `load-path', relative to
;;   this file. Emacs searches the `load-path' when you load packages with
;;   `require' or `use-package'.
;; - `map!' for binding new keys
;;
;; To get information about any of these functions/macros, move the cursor over
;; the highlighted symbol at press 'K' (non-evil users must press 'C-c c k').
;; This will open documentation for it, including demos of how they are used.
;;
;; You can also try 'gd' (or 'C-c c d') to jump to their definition and see how
;; they are implemented.

(prefer-coding-system           'utf-8)
(set-default-coding-systems     'utf-8)
(set-terminal-coding-system     'utf-8)
(set-keyboard-coding-system     'utf-8)
(setq default-buffer-file-encoding-system 'utf-8)
(setq x-select-request-type '(UTF8_STRING COMPOUND_TEXT TEXT STRING))

(setq-default indent-tabs-mode nil)
(setq-default tab-width 4)
(setq c-set-style "k&r")
(setq c-basic-offset 4)

(use-package! tree-sitter
  :config
  (require 'tree-sitter-langs)
  (global-tree-sitter-mode)
  (add-hook 'tree-sitter-after-on-hook #'tree-sitter-hl-mode)
  (pushnew! tree-sitter-major-mode-language-alist
            '(scss-mode . css)))

(defun reo101/org-mode-setup ()
  (org-indent-mode)
  (visual-line-mode 1))

(use-package! org
  :hook (org-mode . reo101/org-mode-setup)
  :custom
  (org-ellipsis " ▼")
  (org-hide-emphasis-markers t)
  (org-src-window-setup 'current-window))

(use-package! org-bullets
  :hook (org-mode . org-bullets-mode)
  :custom (org-bullets-bullet-list '("◉" "○" "●" "○" "●" "○" "●")))

(font-lock-add-keywords 'org-mode
	  '(("^ *\\([-]\\) "
	     (0 (prog1 () (compose-region (match-beginning 1)
                                          (match-end 1) "•"))))))

(defun reo101/org-mode-visual-fill ()
  (setq visual-fill-column-width 125
        visual-fill-column-center-text t)
  (visual-fill-column-mode 1))

(use-package! visual-fill-column
  :diminish visual-line-mode
  :hook (org-mode . reo101/org-mode-visual-fill))

(use-package! highlight-indent-guides
  :commands highlight-indent-guides-mode
  :hook (prog-mode . highlight-indent-guides-mode)
  :config
  (setq highlight-indent-guides-method 'character
        highlight-indent-guides-character ?\»
        highlight-indent-guides-delay 0.01
        highlight-indent-guides-responsive 'top
        highlight-indent-guides-auto-enabled nil)) ;nil

;; (use-package! lsp-mode
;;   :hook ((prog-mode . lsp-deferred))
;;   :commands (lsp lsd-deferred)
;;   :config
;;   (progn
;;     (lsp-register-client
;;      (make-lsp-client :new-connection (lsp-tramp-connection "clangd")
;;                       :major-modes '(c-mode c++-mode)
;;                       :remote? t
;;                       :server-id 'clangd-remote))))

(defun reo101/lsp-mode-setup ()
  (setq lsp-headerline-breadcrumb-segments '(path-up-to-project file symbols))
  (lsp-headerline-breadcrumb-mode))

(use-package! lsp-mode
  :diminish
  :commands (lsp lsp-deferred)
  ;; :bind-keymap ("C-c l" . lsp-command-map)
  :custom
  (lsp-log-io t)
  ;; (lsp-keymap-prefix "C-c l")
  (lsp-register-client
  (make-lsp-client :new-connection (lsp-stdio-connection "intelephense")
                   :major-modes '(c++-mode)
                   :server-id 'intelephense))
  :hook
  ;; (erlang-mode . lsp)
  (c++-mode . lsp)
  ;; (latex-mode . lsp)
  ;; (latex-math-mode . lsp)
  ;; (php-mode . lsp)
  (lsp-mode . reo101/lsp-mode-setup)
  (lsp-mode . lsp-enable-which-key-integration))

(use-package! lsp-ui
  :hook (lsp-mode . lsp-ui-mode)
  :commands lsp-ui
  :custom
  (lsp-ui-sideline-enable t)
  (lsp-ui-doc-enable t)
  (lsp-ui-doc-position 'bottom))

(use-package! lsp-origami
  :hook
  (erlang-mode . origami-mode)
  (origami-mode . lsp-origami-mode))

(use-package! lsp-ivy :commands lsp-ivy-workspace-symbol)

; (use-package! origami
;   :bind-keymap ("C-c o" . origami-mode-map)
;   :bind (:map origami-mode-map
;               ("C-c o o" . origami-open-node)
;               ("C-c o O" . origami-open-node-recursively)
;               ("C-c o c" . origami-close-node)
;               ("C-c o C" . origami-close-node-recursively)
;               ("C-c o a" . origami-toggle-node)
;               ("C-c o A" . origami-recursively-toggle-node)
;               ("C-c o R" . origami-open-all-nodes)
;               ("C-c o M" . origami-close-all-nodes)
;               ("C-c o v" . origami-show-only-node)
;               ("C-c o k" . origami-previous-fold)
;               ("C-c o j" . origami-forward-fold)
;               ("C-c o x" . origami-reset)))

(use-package! company
  :after lsp-mode
  :diminish
  :custom
  (company-idle-delay 0)
  (company-minimum-prefix-length 1)
  :bind (:map company-active-map
              ("M-n" . nil)
              ("M-p" . nil)
              ("C-n" . company-select-next)
              ("C-p" . company-select-previous))
  :hook (lsp-mode . company-mode))

;; (use-package! company-box
;;   :diminish
;;   :hook (company-mode . company-box-mode))

; (defvar my-term-shell "/bin/zsh")
; (defadvice ansi-term (before force-bash)
;   (interactive (list my-term-shell)))
; (ad-activate 'ansi-term)
; (global-set-key (kbd "<M-return>") 'ansi-term)

(use-package! sudo-edit
  :bind ("C-c s" . sudo-edit))

; (use-package! ivy
;   :diminish
;   :bind (:map ivy-minibuffer-map
;               ("TAB" . ivy-alt-done))
;   :config (ivy-mode 1))
; (use-package! ivy-rich
;   :init (ivy-rich-mode 1))

; (set-frame-parameter (selected-frame) 'alpha '(85 . 50))
; (add-to-list 'default-frame-alist '(alpha . (85 . 50)))

(use-package! rainbow-delimiters
  :hook (prog-mode . rainbow-delimiters-mode))

;; (prettify-utils-add-hook tex-mode
;;                          ("\\smallo" "o"))

(use-package! org-super-agenda
  :after org-agenda
  :init
  (setq org-super-agenda-groups '((:name "Today"
                                         :time-grid t
                                         :scheduled today)
                                  (:name "Due today"
                                         :deadline today)
                                  (:name "Important"
                                         :priority "A")
                                  (:name "Overdue"
                                         :deadline past)
                                  (:name "Due soon"
                                         :deadline future)
                                  (:name "Big Outcomes"
                                         :tag "bo")))
  :config
  (org-super-agenda-mode))
